File: app/src/main/java/com/e/hardviews/MainModel.java
package com.e.hardviews;

import java.util.ArrayList;
import java.util.List;

import static com.e.hardviews.Action.BAD_HABITS;
import static com.e.hardviews.Action.FOOD;
import static com.e.hardviews.Action.HEALTH;
import static com.e.hardviews.Action.TIME;

public class MainModel {

    AppDatabase db = App.getInstance().getDatabase();
    MyActionsDBDao actionsDBDao = db.actionsDBDao();

    private final String CREATE_ACTION = "Add a Task";
    private List<Action> actions;
    private final List<Action> defaultActions;

    public MainModel (){
        actions = getAllFromDB();
        defaultActions = createDefaultActions();
    }

    private List<Action> getAllFromDB() {
        new Thread(){
            @Override
            public void run() {
                actions = actionsDBDao.getAll();
            }
        }.start();
        return actions;
    }

    private List<Action> createDefaultActions(){
        List<Action> defaultActionList = new ArrayList<>();
        defaultActionList.add(new Action("brush teeth", R.drawable.ic_tooth, R.drawable.ic_tooth_reverse, 1, HEALTH));
        defaultActionList.add(new Action("walk the dog", R.drawable.ic_dog_side, R.drawable.ic_dog_side_reverse, 1,TIME));
        defaultActionList.add(new Action("Running", R.drawable.ic_run_fast_reverse, R.drawable.ic_run_fast, 1,HEALTH));
        defaultActionList.add(new Action("Take a vitamins", R.drawable.ic_pill_reverse, R.drawable.ic_pill, 1,HEALTH));
        defaultActionList.add(new Action("Homework", R.drawable.ic_lead_pencil_reverse, R.drawable.ic_lead_pencil, 1,TIME));
        defaultActionList.add(new Action("Read a book", R.drawable.ic_bookshelf_reverse, R.drawable.ic_bookshelf, 1,TIME));
        defaultActionList.add(new Action("Cook a chiken", R.drawable.ic_food_turkey_reverse, R.drawable.ic_food_turkey, 1,FOOD));
        defaultActionList.add(new Action("Take a food", R.drawable.ic_pasta_reverse, R.drawable.ic_pasta, 1,FOOD));
        defaultActionList.add(new Action("No alcohol", R.drawable.ic_glass_cocktail_reverse, R.drawable.ic_glass_cocktail, 1,BAD_HABITS));
        defaultActionList.add(new Action("No smoking", R.drawable.ic_smoking_off_reverse, R.drawable.ic_smoking_off, 1,BAD_HABITS));
        return defaultActionList;
    }

    private void insertIntoDB(final Action someAction){
        new Thread(){
            @Override
            public void run() {
                actionsDBDao.insert(someAction);
            }
        }.start();
    }

    private void deleteFromDB(final Action chosenAction){
        new Thread(){
            @Override
            public void run() {
                actionsDBDao.delete(chosenAction);

            }
        }.start();
    }

    public List<Action> getActions() {
        return actions;
    }

    public List<Action> getDefaultActions() {
        return defaultActions;
    }

    public void addActionForCreate(){
        boolean checkCreator = false;
        for (int i = 0; i < actions.size(); i++){
            Action action = actions.get(i);
            if (action.isCreator()) checkCreator = true;
        }
        if (!checkCreator){
            Action creator = new Action(CREATE_ACTION, R.drawable.ic_plus_thick, R.drawable.ic_plus_thick, 1, 1);
            creator.setCreator(true);
            actions.add(creator);
        }
    }

    public void deleteActionForCreate(){
        for (int i = 0; i < actions.size(); i++){
            Action action = actions.get(i);
            if (action.getNameAction().equals(CREATE_ACTION)){
                actions.remove(action);
            }
        }
    }

    public void addNewAction(Action newAction){
        insertIntoDB(newAction);
        actions = getAllFromDB();
    }

    public void editAction(Action chosenAction){
       insertIntoDB(chosenAction);
        actions = getAllFromDB();
    }

    public void deleteAction(Action action){
        deleteFromDB(action);
        actions = getAllFromDB();
    }
}

File: app/src/main/java/com/e/hardviews/MainModel.kt
package com.e.hardviews

import android.util.Log
import androidx.lifecycle.LiveData
import kotlinx.coroutines.*
import java.util.*

class MainModel {

    private var db = App.getInstance().database
    private var actionsDBDao = db.actionsDBDao()
    private var defaultActions: List<Action>
    private val ioScope = CoroutineScope(Job() + Dispatchers.IO
            + CoroutineExceptionHandler { coroutineContext, throwable ->
        Log.e("coroutineTag", throwable.message ?: "")
    })

    init { defaultActions = createDefaultActions() }

    private fun createDefaultActions(): List<Action> {
        val defaultActionList: MutableList<Action> = ArrayList()
        defaultActionList.add(Action("brush teeth", R.drawable.ic_tooth, R.drawable.ic_tooth_reverse, 1, Action.HEALTH))
        defaultActionList.add(Action("walk the dog", R.drawable.ic_dog_side, R.drawable.ic_dog_side_reverse, 1, Action.TIME))
        defaultActionList.add(Action("Running", R.drawable.ic_run_fast_reverse, R.drawable.ic_run_fast, 1, Action.HEALTH))
        defaultActionList.add(Action("Take a vitamins", R.drawable.ic_pill_reverse, R.drawable.ic_pill, 1, Action.HEALTH))
        defaultActionList.add(Action("Homework", R.drawable.ic_lead_pencil_reverse, R.drawable.ic_lead_pencil, 1, Action.TIME))
        defaultActionList.add(Action("Read a book", R.drawable.ic_bookshelf_reverse, R.drawable.ic_bookshelf, 1, Action.TIME))
        defaultActionList.add(Action("Cook a chiken", R.drawable.ic_food_turkey_reverse, R.drawable.ic_food_turkey, 1, Action.FOOD))
        defaultActionList.add(Action("Take a food", R.drawable.ic_pasta_reverse, R.drawable.ic_pasta, 1, Action.FOOD))
        defaultActionList.add(Action("No alcohol", R.drawable.ic_glass_cocktail_reverse, R.drawable.ic_glass_cocktail, 1, Action.BAD_HABITS))
        defaultActionList.add(Action("No smoking", R.drawable.ic_smoking_off_reverse, R.drawable.ic_smoking_off, 1, Action.BAD_HABITS))
        return defaultActionList
    }

    fun getAllFromDB(): LiveData<List<Action>> = actionsDBDao.all

    fun getDefaultActions(): List<Action> = defaultActions

    fun addNewAction(newAction: Action) {
        ioScope.launch { actionsDBDao.insert(newAction) }
    }

    fun editAction(chosenAction: Action) {
        ioScope.launch { actionsDBDao.insert(chosenAction) }
    }

    fun deleteAction(action: Action) {
        ioScope.launch { actionsDBDao.delete(action) }
    }
}